tensorflow/contrib/correlation_cost/kernels/correlation_cost_op_gpu.cu.cc

```diff
@@ -0,0 +1,449 @@
+      continue;
+    }
+
+    if (Wmin > Wmax || Hmin > Hmax) {
+      // assumes gradInput2 is pre-allocated and zero filled
+      continue;
+    }
+
+    Wmin = max(0, Wmin);
+    Wmax = min(Wout - 1, Wmax);
+
+    Hmin = max(0, Hmin);
+    Hmax = min(Hout - 1, Hmax);
+
+    const int indx1 =
+        n * (pHin * pWin * Cin) + (h - j2) * (pWin * Cin) + (w - i2) * Cin + c;
+    const float val1 = rInput1[indx1];
+
+    for (int j = Hmin; j <= Hmax; ++j) {
+      for (int i = Wmin; i <= Wmax; ++i) {
+        const int tindx =
+            n * (Cout * Hout * Wout) + tc * (Hout * Wout) + j * Wout + i;
+        thread_accumulation += gradOutput[tindx] * val1;
+      }
+    }
+  }
+  __syncthreads();
+
+  const float reduce_sum =
+      WarpReduce(temp_sum_storage).Sum(thread_accumulation);
+  if (t0 == 0) {
+    if (is_NCHW) {
+      const int indx2 = n * (Cin * Hin * Win) + c * (Hin * Win) +
+                        (h - pad_size) * (Win) + (w - pad_size);
+      gradInput2[indx2] = reduce_sum / nelems;
+    } else {
+      const int indx2 = n * (Cin * Hin * Win) + (h - pad_size) * (Win * Cin) +
+                        (w - pad_size) * Cin + c;
+      gradInput2[indx2] = reduce_sum / nelems;
+    }
+  }
+}
+
+};  // namespace
+
+template <typename Dtype>
+struct CorrelationCostFunctor<GPUDevice, Dtype> {
+  Status operator()(OpKernelContext *context, const Tensor &input_a_t,
+                    const Tensor &input_b_t, Tensor *padded_a_t,
+                    Tensor *padded_b_t, Tensor *output_t,
+                    /* params */
+                    int kernel_size, int max_displacement, int stride_1,
+                    int stride_2, int pad, TensorFormat data_format) {
+    // do not change: the CUDA kernels expects THREADS_PER_BLOCK==32
+    const int THREADS_PER_BLOCK = 32;
+
+    const int32 B = GetTensorDim(input_a_t, data_format, 'N');
+    const int32 iC = GetTensorDim(input_a_t, data_format, 'C');
+    const int32 iH = GetTensorDim(input_a_t, data_format, 'H');
+    const int32 iW = GetTensorDim(input_a_t, data_format, 'W');
+
+    dim3 blocks_grid(B, iH, iW);
+    dim3 threads_block(THREADS_PER_BLOCK);
+
+    // the output is always NCHW (python transposes it to NHWC)
+    const int32 oC = GetTensorDim(*output_t, FORMAT_NCHW, 'C');
+    const int32 oH = GetTensorDim(*output_t, FORMAT_NCHW, 'H');
+    const int32 oW = GetTensorDim(*output_t, FORMAT_NCHW, 'W');
+
+    // set everything to zero (we zero-pad)
+    cudaMemset(padded_a_t->flat<Dtype>().data(), 0,
+               padded_a_t->NumElements() * sizeof(Dtype));
+    cudaMemset(padded_b_t->flat<Dtype>().data(), 0,
+               padded_b_t->NumElements() * sizeof(Dtype));
+    cudaMemset(output_t->flat<Dtype>().data(), 0,
+               output_t->NumElements() * sizeof(Dtype));
+
+    const bool is_NCHW = (data_format == FORMAT_NCHW);
+    if (is_NCHW) {
+      pad_and_transpose<THREADS_PER_BLOCK><<<blocks_grid, threads_block>>>(
```

Comment:
This could be `tf.pad(tf.tranpose(...))` in `ops/correlation_cost_op.py` which is easier to maintain . But the CPU version does not use padded inputs anyway and it would be an additional kernel call.